user-operator/src/main/java/io/strimzi/operator/user/operator/ScramShaCredentials.java

```diff
@@ -4,199 +4,176 @@
-            throw new RuntimeException("Given number of iterations (" + iterations + ") " +
-                    "is less than minimum iterations for mechanism (" + mechanism.minIterations() + ")");
-        }
-        StringBuilder value = new StringBuilder(mechanism.mechanismName()).append("=[");
-        if (iterations > 0) {
-            value.append("iterations=").append(iterations).append(',');
-        }
-        value.append("password=").append(password).append(']');
-        try (ProcessHelper.ProcessResult pr = exec(asList(
-                    "--zookeeper", zookeeper,
-                    "--alter",
-                    "--entity-name", username,
-                    "--entity-type", "users",
-                    "--add-config", value.toString()))) {
-            Pattern compile = Pattern.compile("Completed Updating config for entity: user-principal '.*'\\.");
-            if (!matchResult(pr, pr.standardOutput(), 0, compile)) {
-                throw unexpectedOutput(pr);
-            }
+    public void createOrUpdate(String username, String password) {
+        byte[] data = zkClient.readData("/config/users/" + username, true);
+
+        if (data != null)   {
+            log.debug("Updating {} credentials for user {}", mechanism.mechanismName(), username);
+            zkClient.writeData("/config/users/" + username, encodeUser(password));
+        } else {
+            log.debug("Creating {} credentials for user {}", mechanism.mechanismName(), username);
+            ensurePath("/config/users");
+            zkClient.createPersistent("/config/users/" + username, encodeUser(password));
         }
+
+        notifyChanges(username);
     }
 
     /**
      * Delete the SCRAM-SHA credentials for the given user.
      * It is not an error if the user doesn't exist, or doesn't currently have any SCRAM-SHA credentials.
+     *
+     * @param username Name of the user
      */
     public void delete(String username) {
-        try (ProcessHelper.ProcessResult pr = exec(asList(
-                "--zookeeper", zookeeper,
-                "--alter",
-                "--entity-name", username,
-                "--entity-type", "users",
-                "--delete-config", mechanism.mechanismName()))) {
-            if (!matchResult(pr, pr.standardOutput(), 0,
-                    Pattern.compile("Completed Updating config for entity: user-principal '.*'\\."))
-                    && !matchResult(pr, pr.standardError(), 1,
-                    Pattern.compile(Pattern.quote("Invalid config(s): " + mechanism.mechanismName())))) {
-                throw unexpectedOutput(pr);
-            }
+        byte[] data = zkClient.readData("/config/users/" + username, true);
+
+        if (data != null)   {
+            log.debug("Deleting {} credentials for user {}", mechanism.mechanismName(), username);
+
+            JsonObject json = new JsonObject()
+                    .put("version", 1)
+                    .put("config", new JsonObject());
```

Comment:
Users can have other configs apart from `SCRAM-SHA-512`: `producer_byte_rate`, `consumer_byte_rate`, `request_percentage`. To preserve those we should really fetch their existing configs, remove `SCRAM-SHA-512` and update.

Reply:
Fair enough ... I didn't know that quotas are stored here as well. I thought it's only the SCRAM credentials.